### PR TITLE
Async loading with new ChunkLoader class

### PR DESCRIPTION
**Pull request by pwinston**
This is only the ChunkLoader parts of https://github.com/napari/napari/pull/1354, see that description.

Once the perfmon PR https://github.com/napari/napari/pull/1453 is merged we'll open up a PR for this branch `feature/chunk-loader2` and give it a new description, etc.

### CODE REVIEW

**Comment by pwinston** [comment]
Add config file:

1. If `NAPARI_ASYNC`is not defined or is "0" you get synchronous.
2. If `NAPARI_ASYNC` is "1" you get asynchronous with defaults.
3. Otherwise we assume `NAPARI_ASYNC` is the path to a config file.

So far the config file looks like this:

```
{
    "synchronous": false,
    "num_workers": 6,
    "use_processes": true,
    "log_path": "/Users/pbw/dev/napari/chunk.log"
}
```

**Comment by d-v-b** [comment]
I tried the following example:
```python
import os
os.environ['NAPARI_ASYNC'] = '1'
import napari
import dask.array as da
data = da.random.randint(0, 255, (100,) * 3, chunks=(10,) * 3, dtype='uint8')
data_computed = data.compute()
with napari.gui_qt():
    v = napari.Viewer()
    v.add_image(data)
    v.add_image(data_computed)
```

`data_computed` was visualized fine, but `data` failed to render, and napari emitted a lot of exceptions based on attribute errors:

```
exception calling callback for <Future at 0x7fa1885508e0 state=finished returned ChunkRequest>
Traceback (most recent call last):
  File "/Users/bennettd/miniconda3/envs/napari/lib/python3.8/concurrent/futures/_base.py", line 328, in _invoke_callbacks
    callback(self)
  File "/Users/bennettd/dev/napari/napari/utils/chunk/_loader.py", line 245, in _done
    request.add_perf_event()
  File "/Users/bennettd/dev/napari/napari/utils/chunk/_request.py", line 79, in add_perf_event
    timers.add_event(
AttributeError: 'NoneType' object has no attribute 'add_event'
```

Is my example poorly chosen?

**Comment by pwinston** [comment]
Thanks for trying it. I pushed a fix for this. It was crashing because I had not tested things with perfmon disabled recently. In this one spot we were mistakenly assuming perfmon was on. Which it may or may not be. Fixed now.

**Comment by d-v-b** [comment]
No more crashes! I amended my example to include an image where the chunks should take a variable amount of time to "load":
```python
import os
os.environ['NAPARI_ASYNC'] = '1'
import napari
import dask.array as da
import time
import random

data = da.random.randint(0, 255, (100,) * 3, chunks=(10,) * 3, dtype='uint8')

def func(v):
    # return input after a delay
    time.sleep(random.random())
    return v

data2 = data.map_blocks(func)
data_computed = data.compute()

with napari.gui_qt():
    v = napari.Viewer()
    v.add_image(data)
    v.add_image(data2)
    v.add_image(data_computed)

```
When visualizing `data2` (where each chunk should arrive after a random delay) nothing renders until all the chunks have arrived -- is this the desired behavior for this PR?

**Comment by pwinston** [comment]
I've not used `map_blocks` before but I don't think it works how you are hoping it works. When I run that in napari it takes about 5 seconds before showing each slice. And we're doing 1 chunk load, but that 1 load takes 5 seconds.

Which seems very weird, why 5 seconds? It would take about 50 seconds to load the whole thing. And then if go to the very next slice it take another 5 seconds.

I'm not sure I can totally explain it but see this gist:
https://gist.github.com/pwinston/39fa2fc420f9d521964c6738b32462ec

And see the output from running it:
https://gist.github.com/pwinston/9fb6f0263534f51be55749b79e9d258c

Basically data2 is such that if you ask for 1 slice it will compute the entire thing using 12 threads and it will take another 5 seconds. So that's exactly what we're seeing in napari, napari is taking 5 seconds to pull out each slice. So I think that's just how map_blocks works in this case?

**Comment by d-v-b** [comment]
`array.map_blocks` just allows you to create a new array by mapping some function over the chunks in your old array. So `data2` is exactly the same values as `data`, but each 10x10x10 chunk arrives after a random delay. If each chunk was rendered as soon as it arrives, I would see little squares of data popping into existence. But instead I see all the data (in a slice) arriving at once. So I'm just wondering if this is the desired behavior at this stage of the async rendering PR.

**Comment by d-v-b** [comment]
>When I run that in napari it takes about 5 seconds before showing each slice. And we're doing 1 chunk load, but that 1 load takes 5 seconds.

A key detail here is that chunks are 10x10x10, so a single slice along z is 100 independent chunks.

**Comment by pwinston** [comment]
The PR adds async loading for single-scale images, not multi-scale images. So there is no way for a "portion of a single slice" to arrive before any other portion. One entire slice is requested at a time from the ArrayLike interview so one entire slice will arrive at a time.



Separate from that though, the `map_blocks` here seems strange to me. When we ask for a slice, in my test program not even in napari, it uses 12 threads over a period of 5 seconds to compute the slice. But when we ask for an adjacent slice it appears to do the full computation over again.

That seems strange to me. If it was lazily computing the final array you'd expect asking for the adjacent slice would take no time, because the 10x10x10 chunks computed for the full slice would represent the final results of 10 adjacent slices, wouldn't it? So you'd expect asking for the next 9 slices would take zero time, and only when you crossed the chunk boundary would it re-compute anything.

So this PR is only for single-scale, but even if it were multi-scale it seems like `map_blocks` here would compute the same data over and over, wouldn't?

**Comment by d-v-b** [comment]
My example is not multiscale -- It is a single array divided into cubic chunks, which is super common for large image volumes stored in zarr / n5 format. In these datasets, a single slice could be 10000x10000, which is exactly why you _don't_ want to wait for a single slice to load before drawing. Instead, tools like neuroglancer of bigdataviewer draw each chunk as it arrives.

>That seems strange to me. If it was lazily computing the final array you'd expect asking for the adjacent slice would take no time, because the 10x10x10 chunks computed for the full slice would represent the final results of 10 adjacent slices, wouldn't it? So you'd expect asking for the next 9 slices would take zero time, and only when you crossed the chunk boundary would it re-compute anything.

Dask does not cache by default, so if you call `a = data2[0].compute(); b = data2[1].compute()` dask computes the exact same thing twice, then takes different slices of the results.

**Comment by pwinston** [comment]
This PR corresponds to https://github.com/napari/napari/pull/1354 which as it says it to fix https://github.com/napari/napari/issues/1320 which deals with single-scale images.

If you run your above example in master you'll see the UI blocks for about 5 seconds as you drag the slider because it's doing the loading/computation/sleeps in the GUI thread. With this PR the UI no longer blocks because it's loading the chunk (one slice of one layer) in a worker thread (or process if you configure it that way). So that's what the PR implements.

I do see a problem though. In master I see the "10 slices get loaded at a time" behavior I was expecting in the first place. I guess because napari turns on the dask cache. But in this PR the caching seems to be gone. So it loads in a worker thread, which is good, but every slice seems to take ~5 seconds as if the cache is off. I don't know what's going on there, we don't alter the cache settings. That needs to be fixed though.

What you were hoping for, that your single-scale image layer would load in chunks, is https://github.com/napari/napari/issues/1300 that we plan to tackle nest along with https://github.com/napari/napari/issues/845. The big change those PR's will be introducing an octree that divides the image into chunks. In https://github.com/napari/napari/issues/1300 there are no underlying chunks but we'll impose them, we'll be able to ask for 256x256 tiles and that tile will load and draw independently of the other tiles.

In https://github.com/napari/napari/issues/845 there are underlying chunks because the file format is chunked. Napari's chunks may or many correspond to the disk chunks, it's up to use to choose the best chunks size automatically or let the user configure it.

**Comment by d-v-b** [comment]
OK! That makes sense. I was premature in hoping to see tiny chunks popping to existence.

**Comment by jni** [comment]
@pwinston "Brave" tester number 2 here! 👋 😂 

I've run this branch with https://github.com/jni/napariboard-proto/blob/master/n2s-mnist-napariboard.ipynb (a cleaned up version of the original motivating example, which removes the issues with dask.delayed on the input volumes), and it's unfortunately still slow to switch slices. Is that expected? e.g. compare scrolling through the stack with the output layer visible vs invisible. The discussion is spread over a few issues so I don't know what the current expected state is. Is this the issue you were running into with dask threads competing with Qt threads? Also, are we doomed? =P

**Comment by pwinston** [comment]
Closed now that we opened https://github.com/napari/napari/pull/1495 in the napari repo.
